mods/d2k/maps/harkonnen-09a/harkonnen09a.lua

```diff
@@ -0,0 +1,281 @@
+		corrino_small.MarkCompletedObjective(KillHarkonnen5)
+	end
+
+	if atreides_main.HasNoRequiredUnits() and atreides_small_1.HasNoRequiredUnits() and atreides_small_2.HasNoRequiredUnits() and not player.IsObjectiveCompleted(KillAtreides) then
+		Media.DisplayMessage("The Atreides have been annihilated!", "Mentat")
+		player.MarkCompletedObjective(KillAtreides)
+	end
+
+	if corrino_main.HasNoRequiredUnits() and corrino_small.HasNoRequiredUnits() and not player.IsObjectiveCompleted(KillCorrino) then
+		Media.DisplayMessage("The Emperor has been annihilated!", "Mentat")
+		player.MarkCompletedObjective(KillCorrino)
+	end
+
+	if DateTime.GameTime % DateTime.Seconds(10) == 0 and LastHarvesterEaten[atreides_main] then
+		local units = atreides_main.GetActorsByType("harvester")
+
+		if #units > 0 then
+			LastHarvesterEaten[atreides_main] = false
+			ProtectHarvester(units[1], atreides_main, AttackGroupSize[Difficulty])
+		end
+	end
+
+	if DateTime.GameTime % DateTime.Seconds(10) == 0 and LastHarvesterEaten[atreides_small_1] then
+		local units = atreides_small_1.GetActorsByType("harvester")
+
+		if #units > 0 then
+			LastHarvesterEaten[atreides_small_1] = false
+			ProtectHarvester(units[1], atreides_small_1, AttackGroupSize[Difficulty])
+		end
+	end
+
+	if DateTime.GameTime % DateTime.Seconds(10) == 0 and LastHarvesterEaten[corrino_main] then
+		local units = corrino_main.GetActorsByType("harvester")
+
+		if #units > 0 then
+			LastHarvesterEaten[corrino_main] = false
+			ProtectHarvester(units[1], corrino_main, AttackGroupSize[Difficulty])
+		end
+	end
+
+	if DateTime.GameTime % DateTime.Seconds(10) == 0 and LastHarvesterEaten[corrino_small] then
+		local units = corrino_small.GetActorsByType("harvester")
+
+		if #units > 0 then
+			LastHarvesterEaten[corrino_small] = false
+			ProtectHarvester(units[1], corrino_small, AttackGroupSize[Difficulty])
+		end
+	end
+end
+
+WorldLoaded = function()
+	atreides_main = Player.GetPlayer("Atreides Main Base")
+	atreides_small_1 = Player.GetPlayer("Atreides Small Base 1")
+	atreides_small_2 = Player.GetPlayer("Atreides Small Base 2")
+	corrino_main = Player.GetPlayer("Corrino Main Base")
+	corrino_small = Player.GetPlayer("Corrino Small Base")
+	player = Player.GetPlayer("Harkonnen")
+
+	InitObjectives(player)
+	KillAtreides = player.AddPrimaryObjective("Destroy the Atreides.")
+	KillCorrino = player.AddPrimaryObjective("Destroy the Imperial Forces.")
+	KillHarkonnen1 = atreides_main.AddPrimaryObjective("Kill all Harkonnen units.")
+	KillHarkonnen2 = atreides_small_1.AddPrimaryObjective("Kill all Harkonnen units.")
+	KillHarkonnen3 = atreides_small_2.AddPrimaryObjective("Kill all Harkonnen units.")
+	KillHarkonnen4 = corrino_main.AddPrimaryObjective("Kill all Harkonnen units.")
+	KillHarkonnen5 = corrino_small.AddPrimaryObjective("Kill all Harkonnen units.")
+
+	Camera.Position = HMCV.CenterPosition
+	AtreidesAttackLocation = HarkonnenRally.Location
+
+	Trigger.OnAllKilledOrCaptured(AtreidesMainBase, function()
```

Comment:
Atreides would be defeated if all structures are lost, right? Because then this Trigger is unnecessary. It's a bit sad that there's no `OnAnyKilledOrCaptured` (there is only `OnAnyKilled`).

Reply:
No, they lose when all units and buildings are lost. This is here so AI will do a last attack with all its units after its buildings are destroyed.

Reply:
Ahh! Ok. Thanks for the explanation.